src/v2/git_url.rs

```diff
@@ -71,6 +71,44 @@ impl GitUrl {
             }
         }
     }
+
+    /// Returns a new GitUrl which is the same as the
+    /// this one, but without any subdirectory part
+    pub fn without_subdirectory(&self) -> GitUrl {
+        let (repository, branch, _) = self.parse_parts();
+        let branch_str = match branch {
+            Some(branch) => format!("#{}", branch),
+            None => String::new(),
+        };
+        GitUrl { url: format!("{}{}", repository, branch_str) }
+    }
+
+    /// Extract the repository part of the URL
+    pub fn repository(&self) -> String {
+        self.parse_parts().0
+    }
+
+    /// Extract the optional branch part of the git URL
+    pub fn branch(&self) -> Option<String> {
+        self.parse_parts().1
+    }
+
+    /// Extract the optional subdirectory part of the git URL
+    pub fn subdirectory(&self) -> Option<String> {
+        self.parse_parts().2
+    }
+
+    fn parse_parts(&self) -> (String, Option<String>, Option<String>) {
+        lazy_static! {
+            static ref URL_PARSE: Regex = Regex::new(r#"^([^#]+)(#([^:]+)?(:(.+))?)?$"#).unwrap();
+        }
+        let captures = URL_PARSE.captures(&self.url).unwrap();
+        (
+            captures.at(1).unwrap().to_string(),
+            captures.at(3).map(|branch| branch.to_string()),
+            captures.at(5).map(|branch| branch.to_string()),
```

Comment:
Since we're using `&str` everywhere, we can delete the `.to_string()` calls here and save a bunch of heap allocations.

Reply:
Oh wow yeah that's way cleaner. Not sure why I was using Strings everywhere.